subprojects/files/src/main/java/org/gradle/api/internal/file/collections/FileTreeAdapter.java

```diff
@@ -129,6 +129,9 @@ public void visitLeafCollections(FileCollectionLeafVisitor visitor) {
         } else if (tree instanceof SingletonFileTree) {
             SingletonFileTree singletonFileTree = (SingletonFileTree) tree;
             visitor.visitCollection(ImmutableFileCollection.of(singletonFileTree.getFile()));
+        } else if (tree instanceof ArchiveFileTree) {
+            ArchiveFileTree archiveFileTree = (ArchiveFileTree) tree;
+            visitor.visitCollection(ImmutableFileCollection.of(archiveFileTree.getBackingFile()));
```

Comment:
I think we need to do a `null` check here, right?

Reply:
added